File: fastapi-python-backend/src/middleware/request_context.py
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.requests import Request
from starlette.responses import Response

from ..observability.tracing import emit_request_trace_span, start_request_trace


class RequestContextMiddleware(BaseHTTPMiddleware):
    async def dispatch(self, request: Request, call_next) -> Response:
        trace_context = start_request_trace(request.headers.get('x-trace-id'))

        request.state.trace_id = trace_context.trace_id
        request.state.span_id = trace_context.span_id
        request.state.request_start_time = trace_context.request_start_time

        response = await call_next(request)

        response.headers['x-trace-id'] = trace_context.trace_id
        response.headers['x-span-id'] = trace_context.span_id

        emit_request_trace_span(
            request_id=getattr(request.state, 'request_id', None),
            trace_id=trace_context.trace_id,
            span_id=trace_context.span_id,
            method=request.method,
            path=request.url.path,
            status_code=response.status_code,
            request_start_time=trace_context.request_start_time,
        )

        return response

File: hono-ts-bun-backend/src/middleware/request-context.ts
import type { MiddlewareHandler } from "hono";
import { logger } from "../observability/logger";
import { observeRequest } from "../observability/metrics";
import { emitRequestTraceSpan, startRequestTrace } from "../observability/tracing";
import type { AppVariables } from "../types/context";

export const requestContextMiddleware: MiddlewareHandler<{ Variables: AppVariables }> = async (
  c,
  next,
) => {
  const trace = startRequestTrace(c.req.header("x-trace-id"));
  const existing = c.get("requestContext");

  c.set("requestContext", {
    requestId: existing?.requestId ?? "unknown",
    traceId: trace.traceId,
    spanId: trace.spanId,
    requestStartTime: trace.requestStartTime,
    metricsStartTimeNs: process.hrtime.bigint(),
  });

  c.header("x-trace-id", trace.traceId);
  c.header("x-span-id", trace.spanId);

  await next();

  const context = c.get("requestContext");
  const durationSeconds =
    Number(process.hrtime.bigint() - context.metricsStartTimeNs) / 1_000_000_000;

  observeRequest({
    method: c.req.method,
    route: new URL(c.req.url).pathname,
    statusCode: c.res.status,
    durationSeconds,
  });

  logger.info(
    {
      requestId: context.requestId,
      traceId: context.traceId,
      spanId: context.spanId,
      method: c.req.method,
      path: new URL(c.req.url).pathname,
      statusCode: c.res.status,
      durationMs: Math.round(durationSeconds * 1000),
    },
    "Request completed",
  );

  emitRequestTraceSpan({
    requestId: context.requestId,
    traceId: context.traceId,
    spanId: context.spanId,
    method: c.req.method,
    path: new URL(c.req.url).pathname,
    statusCode: c.res.status,
    requestStartTime: context.requestStartTime,
  });
};
